versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksDBConditionVisitor.java

```diff
@@ -0,0 +1,170 @@
+  /**
+   * This provides a separation of queries on @{ExpressionFunction} from the object itself.
+   * This uses the Visitor design pattern to retrieve object attributes.
+   */
+  private static class RocksDBStrValueVisitor implements ValueVisitor<String> {
+    @Override
+    public String visit(Value value) {
+      return toRocksDBString(value.getValue());
+    }
+
+    @Override
+    public String visit(ExpressionFunction value) {
+      final ExpressionFunction.FunctionName name = value.getName();
+      final List<Value> arguments = value.getArguments();
+      if (arguments.size() != name.getArgCount()) {
+        throw new InvalidParameterException(
+          String.format("Number of arguments provided [%d] does not match the number expected [%d] for %s.",
+            arguments.size(), name.getArgCount(), name));
+      }
+
+      switch (name) {
+        case EQUALS:
+          // Special case SIZE, as the object representation is not contained in one level of ExpressionFunction.
+          if (isSize(arguments.get(0))) {
+            return String.format("%s,%s,%s", Function.SIZE,
+              arguments.get(0).getFunction().getArguments().get(0).accept(this), arguments.get(1).accept(this));
+          }
+
+          return String.format("%s,%s,%s", Function.EQUALS, arguments.get(0).accept(this), arguments.get(1).accept(this));
+        default:
+          throw new UnsupportedOperationException(String.format("%s is not a supported top-level RocksDB function.", name));
+      }
+    }
+
+    @Override
+    public String visit(ExpressionPath value) {
+      return value.getRoot().accept(RocksDBPathVisitor.INSTANCE_NO_QUOTE, true);
+    }
+
+    private boolean isSize(Value value) {
+      return (value.getType() == Value.Type.FUNCTION) && (((ExpressionFunction)value).getName() == ExpressionFunction.FunctionName.SIZE);
+    }
+  }
+
+  /**
+   * This provides a separation of queries on @{ExpressionFunction} from the object itself.
+   * This uses the Visitor design pattern to retrieve object attributes.
+   */
+  private static class RocksDBFunctionHolderValueVisitor implements ValueVisitor<Function> {
+    @Override
+    public Function visit(Value value) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Function visit(ExpressionFunction value) {
+      final ExpressionFunction.FunctionName name = value.getName();
+      final List<Value> arguments = value.getArguments();
+      if (arguments.size() != name.getArgCount()) {
+        throw new InvalidParameterException(
+          String.format("Number of arguments provided [%d] does not match the number expected [%d] for %s.",
+            arguments.size(), name.getArgCount(), name));
+      }
+
+      switch (name) {
+        case EQUALS:
+          // Special case SIZE, as the object representation is not contained in one level of ExpressionFunction.
+          if (isSize(arguments.get(0))) {
+            return new Function(Function.SIZE,
```

Comment:
Why are we converting children into strings? We then end up string parsing during evaluation, I think this needs to be re-evaluated, we're currently using string as an intermediary representation instead of using a proper object representation.

Reply:
This is refactored now to store ExpressionPath in Function. I then drill down into that as needed in the evaluate method.
I'm reluctant to leach more ExpressionPath/ConditionExpression objects into this design so I can socialize it as a replacement for those DynamoDB specific implementations.